### PR TITLE
fix: actually pass search filters to experiment actions

### PR DESCRIPTION
**Pull request by ashtonG**
## Description

This fixes an issue where bulk experiment actions did not work when using select all. The endpoint used a different, incompatible filter format that did not apply the proper filters to the selection. We introduce the new filter as a new parameter (for backwards compatibility's sake) and apply it to all actions.

### one
* On an experiment list page, select 2 experiments, click the actions menu and then click "archive"
  *  the selected experiments should be archived (if the current view does not show archived experiments, they should be hidden)
* in the filters menu, select "Show archived", then click the selection column header, then click "select all"
* click the actions menu and select "unarchive"
* confirm the action
* The experiment list should show the same experiments regardless of the "show archived" filter state

### two
* submit an unmanaged experiment to a project with 20 or more experiments in it with a name beginning with the letter a
* to the same project, submit a managed experiment with a name beginning with the letter z
* on an experiment list of that project, disable infinite pagination
* on page one of the experiment list, select the unmanaged experiment
* on the last page of the experiment list, select the managed experiment
* The ui should show that you have two experiments selected
* click the actions menu and select "pause"
* confirm the action
* There should be a popup saying that one experiment was paused

## Checklist

- [x] Changes have been manually QA'd
- [x] User-facing API changes need the "User-facing API Change" label.
- [x] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [x] Licenses should be included for new code which was copied and/or modified from any external code.

## Ticket
ET-33

### CODE REVIEW

File: webui/react/src/components/TableActionBar.tsx [hunk 1]
```diff
@@ -451,31 +489,37 @@ const TableActionBar: React.FC<Props> = ({
...
-        filters={selectAll ? filters : undefined}
+        experimentIds={selectedExperiments.reduce((acc, experiment) => {
+          if (
+            canActionExperiment(ExperimentAction.Move, experiment) &&
+            permissions.canMoveExperiment({ experiment })
+          ) {
+            acc.push(experiment.id);
+          }
+          return acc;
+        }, [] as number[])}
```
Comment (keita-determined) [inline webui/react/src/components/TableActionBar.tsx:500]: i think its safer

```suggestion
        experimentIds={selectedExperiments.reduce((acc: number[], experiment) => {
          if (
            canActionExperiment(ExperimentAction.Move, experiment) &&
            permissions.canMoveExperiment({ experiment })
          ) {
            acc.push(experiment.id);
          }
          return acc;
        }, [])}
```

File: webui/react/src/components/TableActionBar.tsx [hunk 2]
```diff
@@ -451,31 +489,37 @@ const TableActionBar: React.FC<Props> = ({
...
-            }),
-        )}
-        filters={selectAll ? filters : undefined}
+              workspace: { id: experiment.workspaceId },
+            })
+          ) {
+            acc.push(experiment.id);
+          }
+          return acc;
+        }, [] as number[])}
```
Comment (keita-determined) [inline webui/react/src/components/TableActionBar.tsx:517]: same here

File: webui/react/src/components/TableActionBar.tsx
Comment (keita-determined) [inline webui/react/src/components/TableActionBar.tsx:1]: minor bug: user can still see action menu after applying filter. The action menu visibility should be updated based on the number of filtered experience.

In the video I did
- select all experiments
- Apply filter that matches 0 experiments
- still see the action menu and items in the menu are clickable

https://github.com/determined-ai/determined/assets/109113616/28755e2e-fad1-4ce4-bb11-2c411556c582

File: webui/react/src/components/TableActionBar.tsx [hunk 3]
```diff
@@ -148,54 +148,93 @@ const TableActionBar: React.FC<Props> = ({
...
-  const experimentMap = useMemo(() => {
-    return experiments.filter(Loadable.isLoaded).reduce(
-      (acc, experiment) => {
-        acc[experiment.data.experiment.id] = getProjectExperimentForExperimentItem(
-          experiment.data.experiment,
-          project,
-        );
+  const selectedProjectExperimentMap = useMemo(() => {
+    return [...selectedExperimentsMap].reduce(
+      (acc, [id, { experiment }]) => {
```
Comment (keita-determined) [inline webui/react/src/components/TableActionBar.tsx:155]: and we can remove `as Record<number, ProjectExperiment>` in the bottom. 

```suggestion
      (acc: Record<number, ProjectExperiment>, [id, { experiment }]) => {
```

File: webui/react/src/components/TableActionBar.tsx [hunk 4]
```diff
@@ -148,54 +148,93 @@ const TableActionBar: React.FC<Props> = ({
...
+          conjunction: Conjunction.And,
+          kind: FormKind.Group,
+        },
+      };
+    });
+  }, [selection, validFilterSet, project.id]);
 
   const sendBatchActions = useCallback(
     async (action: BatchAction): Promise<BulkActionResult | void> => {
+      let params: BulkActionParams;
```
Comment (keita-determined) [inline webui/react/src/components/TableActionBar.tsx:216]: idk if you like this way, but i think we want `params` to be const. if not preferred, ignore this

```ts
      const params: BulkActionParams = (() => {
        if (selection.type === 'ONLY_IN') {
          // TODO: when the selection spans pages, this check will not work --
          // should probably warn the user and/handle on the backend
          return {
            experimentIds: selectedExperiments.filter((exp) => !exp.unmanaged).map((exp) => exp.id),
          };
        } else {
          return completeFilterSet.match({
            _: () => ({ experimentIds: [] }),
            Loaded: (filterSet) => {
              return { experimentIds: [], searchFilter: JSON.stringify(filterSet) };
            },
          });
        }
      })();
```

**Review by keita-determined — COMMENTED** [review]
Mostly working good! 
Could you add more test plans, especially edge cases?  
Also selections over pagination doesn't work and found a comment about it in the code. this should be mentioned in the test plan to avoid confusion in release party if the fix wouldnt be in this PR

File: webui/react/src/components/TableActionBar.tsx (hunk 4 above)
Comment (ashtonG) [reply webui/react/src/components/TableActionBar.tsx:216]: i'll use a ternary here -- an iife is overkill here because we're not introducing any other variables.

**Comment by ashtonG** [comment]
@keita-determined updated the pr with a fix for the selection spanning across pages and added a test case that should handle most edge cases.

File: master/internal/api_experiment.go [hunk 5]
```diff
@@ -419,32 +420,38 @@ func (a *apiServer) DeleteExperiments(
...
+	results, experiments, err := experiment.DeleteExperiments(ctx, req.ExperimentIds, req.Filters, req.SearchFilter)
 	if err != nil {
 		return nil, err
 	}
 
+	// gather resolved ids
+	var expIDs []int32
+	for _, exp := range experiments {
+		expIDs = append(expIDs, int32(exp.ID))
+	}
```
Comment (corban-beaird) [inline master/internal/api_experiment.go:432]: question (non-blocking): Why is this cast necessary now, when it had been done prior?

File: master/internal/api_experiment.go [hunk 6]
```diff
@@ -966,14 +973,14 @@ func (a *apiServer) ActivateExperiment(
 func (a *apiServer) ActivateExperiments(
 	ctx context.Context, req *apiv1.ActivateExperimentsRequest,
 ) (*apiv1.ActivateExperimentsResponse, error) {
-	results, err := experiment.ActivateExperiments(ctx, req.ExperimentIds, req.Filters)
+	results, err := experiment.ActivateExperiments(ctx, req.ExperimentIds, req.Filters, req.SearchFilter)
```
Comment (corban-beaird) [inline master/internal/api_experiment.go:976]: question (non-blocking): Are we expecting to deprecate req.Filters, potentially shimming it's behavior into req.SearchFilters?

**Review by keita-determined — APPROVED** [review]
web LGTM

**Comment by corban-beaird** [comment]
I'm not entirely clear on why we need to maintain the old filter parameter and struct for these bulk endpoints, do we have other things using these outside of our WebUI?

**Comment by corban-beaird** [comment]
For the scope of this ticket from a triage perspective, I can understand why we'd leave both filter parameters in; however, for the sake of code hygiene, I'd like to come back and re-evaluate how we approach this problem, by at least removing the need for one or the other parameter, and I'd be curious if we could get to a position where we could eliminate the need for specifying the filters entirely (only allowing actions to be done on a per page level).

Additionally, our bulk select feature feels odd. We provide options to select the first 5, 10, 25, or all experiments; however, our page sizes are restricted to 20, 40, and 80. Not a single option allows selecting an entire page. Why would our bulk select not first select everything in the page and then give the object to select everything in the given project with the filters applied.

File: master/internal/experiment/bulk_action.go [hunk 7]
```diff
@@ -404,7 +429,7 @@ func DeleteExperiments(ctx context.Context,
 			validIDs = append(validIDs, check.ID)
 		}
 	}
-	if filters == nil {
+	if filters == nil && searchFilter == nil {
```
Comment (corban-beaird) [inline master/internal/experiment/bulk_action.go:432]: why is searchfilter being checked here and not in the other bulk action functions?

Comment (corban-beaird) [reply master/internal/experiment/bulk_action.go:432]: This is the kind of bug I'm concerned we'd run into by having two filter parameters, where we check one, but not the other.

File: master/internal/experiment/bulk_action.go [hunk 8]
```diff
@@ -182,12 +198,12 @@ func ToAPIResults(results []ExperimentActionResult) []*apiv1.ExperimentActionRes
 
 // ActivateExperiments works on one or many experiments.
 func ActivateExperiments(ctx context.Context,
-	experimentIds []int32, filters *apiv1.BulkExperimentFilters,
+	experimentIds []int32, filters *apiv1.BulkExperimentFilters, searchFilter *string,
 ) ([]ExperimentActionResult, error) {
 	if filters != nil && filters.States == nil {
 		filters.States = []experimentv1.State{experimentv1.State_STATE_PAUSED}
 	}
```
Comment (corban-beaird) [inline master/internal/experiment/bulk_action.go:205]: It looks like work is being done here to correct the filter to only activate paused experiments, I'm not under the impression the same is being done for `SearchFilters`, this seems like a regression.

Comment (corban-beaird) [reply master/internal/experiment/bulk_action.go:205]: https://github.com/determined-ai/determined/assets/25434156/06aa1a61-b931-4fa1-ad70-324acfa837d2

Comment (corban-beaird) [reply master/internal/experiment/bulk_action.go:205]: There are only 106 experiments; however, we get an error that says it failed on 242 experiments.

Comment (corban-beaird) [reply master/internal/experiment/bulk_action.go:205]: This is the downside of having searchFilters being setup as a string like this is, we don't don't how it's structured until later on when the parsing logic is specified

Comment (corban-beaird) [reply master/internal/experiment/bulk_action.go:212]: I added a check to add in a state filter on relevant endpoints when searchFilter is being used & filter is not.

Comment (corban-beaird) [reply master/internal/experiment/bulk_action.go:212]: IMO this is a bandaid solution for a larger problem with how we approach bulk endpoints

File: master/internal/api_experiment.go (hunk 6 above)
Comment (corban-beaird) [reply master/internal/api_experiment.go:976]: I'd prefer we come up with a more maintainable solution, req.Filters is nice because it explicitly defines it's fields & SearchFilters is nice because it gives us greater flexibility and accurately captures filters that go beyond chained ANDs.

That being said req.Filters doesn't give enough flexibility to express more complicated filters (anything beyond equality, and id exclusion) and SearchFilters is difficult to work with prior to parsing.

I had talked with @maxrussell, and a question that came up is why aren't we communicating an exhaustive list of experiment ids. My initial response had been that we don't since the frontend doesn't have the full set of experiment ids if we are working with a select all that goes across multiple pages.

Do we know how often `select all` is being used and then how often those cases exceed 1 page? I'm worried the approach we've used up to this point for bulk actions is designed for a specific edge case, rather than our user's typical usage pattern.

**Comment by ashtonG** [comment]
closing -- we can circle back to this later.
